lib/srv/db/server.go

```diff
@@ -0,0 +1,420 @@
+	// StreamEmitter is a non-blocking audit events emitter.
+	StreamEmitter events.StreamEmitter
+	// TLSConfig is the *tls.Config for this server.
+	TLSConfig *tls.Config
+	// Authorizer is used to authorize requests coming from proxy.
+	Authorizer auth.Authorizer
+	// GetRotation returns the certificate rotation state.
+	GetRotation func(role teleport.Role) (*services.Rotation, error)
+	// Servers contains a list of database servers this service proxies.
+	Servers []services.DatabaseServer
+	// Credentials are credentials to AWS API.
+	Credentials *credentials.Credentials
+	// OnHeartbeat is called after every heartbeat. Used to update process state.
+	OnHeartbeat func(error)
+}
+
+// CheckAndSetDefaults makes sure the configuration has the minimum required
+// to function.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.DataDir == "" {
+		return trace.BadParameter("missing DataDir")
+	}
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing AuthClient")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("missing AccessPoint")
+	}
+	if c.StreamEmitter == nil {
+		return trace.BadParameter("missing StreamEmitter")
+	}
+	if c.TLSConfig == nil {
+		return trace.BadParameter("missing TLSConfig")
+	}
+	if c.Authorizer == nil {
+		return trace.BadParameter("missing Authorizer")
+	}
+	if c.GetRotation == nil {
+		return trace.BadParameter("missing GetRotation")
+	}
+	if len(c.Servers) == 0 {
+		return trace.BadParameter("missing Servers")
+	}
+	if c.Credentials == nil {
+		session, err := awssession.NewSessionWithOptions(awssession.Options{
+			SharedConfigState: awssession.SharedConfigEnable,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		c.Credentials = session.Config.Credentials
+	}
+	return nil
+}
+
+// Server is an application server. It authenticates requests from the web
+// proxy and forwards them to internal applications.
+type Server struct {
+	// Config is the database server configuration.
+	Config
+	// closeContext is used to indicate the server is closing.
+	closeContext context.Context
+	// closeFunc is the cancel function of the close context.
+	closeFunc context.CancelFunc
+	// middleware extracts identity from client certificates.
+	middleware *auth.Middleware
+	// dynamicLabels contains dynamic labels for database servers.
+	dynamicLabels map[string]*labels.Dynamic
+	// heartbeats holds hearbeats for database servers.
+	heartbeats map[string]*srv.Heartbeat
+	// rdsCACerts contains loaded RDS root certificates for required regions.
+	rdsCACerts map[string][]byte
+	// mu protects access to server infos.
+	mu sync.RWMutex
+	// Entry is used for logging.
+	*logrus.Entry
+}
+
+// New returns a new application server.
+func New(ctx context.Context, config Config) (*Server, error) {
+	err := config.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	localCtx, cancel := context.WithCancel(ctx)
```

Comment:
nit: rename `localCtx -> ctx`, so that you don't accidentally use the unwrapped parent context